src/Parallel/LockfreeUnorderedSet.hpp

```diff
@@ -0,0 +1,451 @@
+        continue;
+      }
+      // The entry is empty. Let's try to set it.
+      std::uint64_t current_key_in_slot = EmptySlotValue;
+      if (not entries_[index].compare_exchange_strong(
+              current_key_in_slot, internal_key, std::memory_order_relaxed,
+              std::memory_order_relaxed) and
+          current_key_in_slot != internal_key) {
+        // Another thread just stole this slot from us. Try next slot.
+        continue;
+      }
+      size_.fetch_add(1, std::memory_order_relaxed);
+      // Successful insert. Return.
+      return true;
+    }
+  }
+  return false;
+}
+
+template <class T, std::uint64_t EmptySlotValue, bool TrackBucketSize>
+bool FixedSizeUnorderedSet<T, EmptySlotValue, TrackBucketSize>::erase(
+    const T key) noexcept {
+  const std::uint64_t internal_key = compute_internal_key(key);
+  const std::uint64_t bucket_index = hash(internal_key) * bucket_size;
+  if constexpr (TrackBucketSize) {
+    std::atomic<std::uint64_t>& this_bucket_size = entries_[bucket_index];
+    for (size_t i = 0; i < this_bucket_size.load(std::memory_order_relaxed);
+         ++i) {
+      const size_t index = bucket_index + 1 + i;
+      const std::uint64_t probed_internal_key =
+          entries_[index].load(std::memory_order_relaxed);
+      if (probed_internal_key == internal_key) {
+        std::uint64_t current_key_in_slot = internal_key;
+        if (not entries_[index].compare_exchange_strong(
+                current_key_in_slot, EmptySlotValue, std::memory_order_relaxed,
+                std::memory_order_relaxed) and
+            current_key_in_slot != EmptySlotValue) {
+          // If the CAS failed and the slot value is not the EmptySlotValue,
+          // then we failed to erase. E.g. another thread could have erased this
+          // value.
+          // NOLINTNEXTLINE(readability-simplify-boolean-expr)
+          return false;
+        }
+        size_.fetch_sub(1, std::memory_order_relaxed);
+        // Shrinking the bucket size is extremely difficult because we use the
+        // bucket size as the index of one past the last element in the
+        // bucket, which may be larger than the number of elements in the
+        // bucket. This means we can only shrink the size if we are removing
+        // the last element, but knowing we erased the last element is
+        // difficult in parallel.
+        return true;
+      }
+    }
+  } else {
+    for (size_t i = 0; i < bucket_size; ++i) {
+      const size_t index = bucket_index + i;
+      const std::uint64_t probed_internal_key =
+          entries_[index].load(std::memory_order_relaxed);
+      if (probed_internal_key == internal_key) {
+        std::uint64_t current_key_in_slot = internal_key;
+        if (not entries_[index].compare_exchange_strong(
+                current_key_in_slot, EmptySlotValue, std::memory_order_relaxed,
+                std::memory_order_relaxed) and
+            current_key_in_slot != EmptySlotValue) {
```

Comment:
Similar to the comment in `insert` on updating `size_`, but I think the correct fix here is just to ignore the value stored in `current_key_in_slot`.